packages/devtools_app/lib/src/screens/debugger/span_parser.dart

```diff
@@ -212,44 +200,47 @@ abstract class _Matcher {
 
   bool scan(Grammar grammar, LineScanner scanner, ScopeStack scopeStack);
 
-  List<ScopeSpan> _applyCapture(
+  void _applyCapture(
+    Grammar grammar,
     LineScanner scanner,
     ScopeStack scopeStack,
     Map<String, dynamic>? captures,
     ScopeStackLocation location,
   ) {
-    final spans = <ScopeSpan>[];
     final lastMatch = scanner.lastMatch!;
     final start = lastMatch.start;
     final end = lastMatch.end;
     final matchStartLocation = location;
-    final matchEndLocation = scanner.location;
     if (captures != null) {
-      if (lastMatch.groupCount <= 1) {
-        scopeStack.add(
-          captures['0']['name'],
-          start: matchStartLocation,
-          end: matchEndLocation,
-        );
-      } else {
-        final match = scanner.substring(start, end);
-        for (int i = 1; i <= scanner.lastMatch!.groupCount; ++i) {
-          if (captures.containsKey(i.toString())) {
-            final capture = scanner.lastMatch!.group(i);
-            if (capture == null) {
-              continue;
-            }
-            final startOffset = match.indexOf(capture);
-            scopeStack.add(
-              captures[i.toString()]['name'],
-              start: matchStartLocation.offset(startOffset),
-              end: matchStartLocation.offset(startOffset + capture.length),
-            );
-          }
+      final match = scanner.substring(start, end);
+      for (int i = 0; i <= lastMatch.groupCount; ++i) {
+        // Skip if we don't have a scope or nested patterns for this capture.
+        if (!captures.containsKey(i.toString())) continue;
+
+        final captureText = lastMatch.group(i);
+        if (captureText == null || captureText.isEmpty) continue;
+
+        final startOffset = match.indexOf(captureText);
+        final capture = captures[i.toString()] as Map<String, Object?>;
```

Comment:
Nit: We should either make `captures` a `Map<String, Object?>` or stick with `Map<String, dynamic>`.

Reply:
Any preference? I tend to prefer using `Object?` over `dynamic` because it's harder to accidentally write code that doesn't give analysis errors, but I see there's quite a lot of `dynamic`s already here (I'm happy to make them all `Object?` if that suits you).

FWIW, I don't think it will avoid the cast on this line though, as even as `Map<String, Object?>` the values are still `Object?` and we know they're `Map`s that we need to read.

Reply:
Yeah, I know we wouldn't get rid of the cast, but then we'd at least be consistent in forcing casts and performing null checks. Before null-safety, I always used `dynamic`, but since that drops the requirement for null checks and can introduce runtime errors, I think we should probably try and use `Object?` instead where possible.

I definitely thought `captures` was a `List<Map<String, dynamic>>`, not a `Map<String, dynamic>`. In that case, can we change it to a `Map<String, Map<String, Object?>>` to be explicit?

Reply:
sgtm! I've also pushed a change that changes all `dynamic` to `Object?` and adds any necessary casts as a result, PTAL!

Reply:
LGTM!